### **LIBRERÍAS**

In [1]:
import numpy as np
import pandas as pd
import sqlite3 as sql
import plotly.graph_objs as go ### para gráficos
import plotly.express as px
import a_funciones as fn

In [2]:
conn=sql.connect('data\\db_movies') ### crear cuando no existe el nombre de cd  y para conectarse cuando sí existe.
cur=conn.cursor() ###para funciones que ejecutan sql en base de datos

In [3]:
cur.execute('select name from sqlite_master where type = "table"')
cur.fetchall()

[('ratings',), ('movies',)]

### **--- Ejercicio clase ---**

In [7]:
movies = pd.read_sql("SELECT * from movies", conn)
ratings = pd.read_sql("SELECT * from ratings", conn)

In [11]:
movies

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

In [12]:
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

In [14]:
ratings_count = pd.read_sql("""SELECT COUNT(DISTINCT userID ) AS total_users FROM ratings""", conn) 
ratings_count

total_users
0          610

Hay un total de 610 usuarios registrados en la base de datos

In [15]:
average_rating = pd.read_sql("""SELECT AVG(rating) AS average_rating FROM ratings WHERE movieId = 1""", conn)
average_rating

average_rating
0         3.92093

La película con movieID = 1 Toy Story (1995), tiene un promedio de calificación de 3.92

In [17]:
no_ratings = pd.read_sql("""
    SELECT title FROM movies
    LEFT JOIN ratings ON movies.movieId = ratings.movieId
    WHERE ratings.movieId IS NULL;
""", conn)
no_ratings

title
0                          Innocents, The (1961)
1                                 Niagara (1953)
2                         For All Mankind (1989)
3   Color of Paradise, The (Rang-e khoda) (1999)
4                 I Know Where I'm Going! (1945)
5                             Chosen, The (1981)
6    Road Home, The (Wo de fu qin mu qin) (1999)
7                                 Scrooge (1970)
8                                   Proof (1991)
9                      Parallax View, The (1974)
10                      This Gun for Hire (1942)
11                  Roaring Twenties, The (1939)
12                   Mutiny on the Bounty (1962)
13            In the Realms of the Unreal (2004)
14                      Twentieth Century (1934)
15                     Call Northside 777 (1948)
16                  Browning Version, The (1951)
17                            Chalet Girl (2011)

Las películas mostradas anteriormente, no cuentan con ninguna calificación en la base de datos.

In [18]:
one_rating = pd.read_sql("""
    SELECT movies.title FROM movies
    JOIN ratings ON movies.movieId = ratings.movieId
    GROUP BY movies.movieId, movies.title HAVING COUNT(ratings.rating) = 1;
""", conn)
one_rating

title
0                        When Night Is Falling (1995)
1                                      Georgia (1995)
2                                    Nico Icon (1995)
3     Once Upon a Time... When We Were Colored (1995)
4                       In the Bleak Midwinter (1995)
...                                               ...
3441        Black Butler: Book of the Atlantic (2017)
3442                     No Game No Life: Zero (2017)
3443                                     Flint (2017)
3444              Bungo Stray Dogs: Dead Apple (2018)
3445              Andrew Dice Clay: Dice Rules (1991)

[3446 rows x 1 columns]

Las películas mostradas anteriormente, solo cuentan con una calificación en la base de datos.